okta/policies.go

```diff
@@ -0,0 +1,887 @@
+		SelfServiceUnlock        PasswordAction `json:"selfServiceUnlock,omitempty"`
+	} `json:"actions,omitempty"`
+	Links `json:"_links,omitempty"`
+}
+
+type SignOn struct {
+	Access                  string `json:"access,omitempty"`
+	RequireFactor           bool   `json:"requireFactor,omitempty"`
+	FactorPromptMode        string `json:"factorPromptMode,omitempty"`
+	RememberDeviceByDefault bool   `json:"rememberDeviceByDefault,omitempty"`
+	FactorLifetime          int    `json:"factorLifetime,omitempty"`
+	Session                 struct {
+		MaxSessionIdleMinutes     int  `json:"maxSessionIdleMinutes,omitempty"`
+		MaxSessionLifetimeMinutes int  `json:"maxSessionLifetimeMinutes,omitempty"`
+		UsePersistentCookie       bool `json:"usePersistentCookie,omitempty"`
+	} `json:"session,omitempty"`
+}
+
+// rule actions for passwords use the same passwordAction obj
+type PasswordAction struct {
+	Access string `json:"access,omitempty"`
+}
+
+// GetPolicyRules returns an array of Rule objs
+type rules struct {
+	Rules []Rule `json:",-"`
+}
+
+// Get a policy
+// Requires Policy ID from Policy object
+func (p *PoliciesService) GetPolicy(id string) (*Policy, *Response, error) {
+	u := fmt.Sprintf("policies/%v", id)
+	req, err := p.client.NewRequest("GET", u, nil)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	policy := new(Policy)
+	resp, err := p.client.Do(req, policy)
+	if err != nil {
+		return nil, resp, err
+	}
+
+	return policy, resp, err
+}
+
+// Get all policies by type
+// Allowed types are OKTA_SIGN_ON, PASSWORD, MFA_ENROLL, or OAUTH_AUTHORIZATION_POLICY
+func (p *PoliciesService) GetPolicesByType(policyType string) (*policies, *Response, error) {
+	u := fmt.Sprintf("policies?type=%v", policyType)
+	req, err := p.client.NewRequest("GET", u, nil)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	policy := make([]Policy, 0)
+	resp, err := p.client.Do(req, &policy)
+	if err != nil {
+		return nil, resp, err
+	}
+	if len(policy) > 0 {
+		myPolicies := new(policies)
+		for _, v := range policy {
+			myPolicies.Policies = append(myPolicies.Policies, v)
+		}
+		return myPolicies, resp, err
+	}
+
+	return nil, resp, err
+}
+
+// Delete a policy
+// Requires Policy ID from Policy object
+func (p *PoliciesService) DeletePolicy(id string) (*Response, error) {
+	u := fmt.Sprintf("policies/%v", id)
+	req, err := p.client.NewRequest("DELETE", u, nil)
+	if err != nil {
+		return nil, err
+	}
+	resp, err := p.client.Do(req, nil)
+	if err != nil {
+		return resp, err
+	}
+
+	return resp, err
+}
+
+// Create a policy
+// You must pass in the Policy object created from Policies.InputPolicy()
+func (p *PoliciesService) CreatePolicy(policy interface{}) (*Policy, *Response, error) {
+	u := fmt.Sprintf("policies")
+	req, err := p.client.NewRequest("POST", u, policy)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	newPolicy := new(Policy)
+	resp, err := p.client.Do(req, newPolicy)
+	if err != nil {
+		return nil, resp, err
+	}
```

Comment:
There is quite a bit of repetitive code in here, is there any way we can pull some of the repeats out into a helper method:

Specifically this seems to repeat a lot:
```
+	if err != nil {
+		return nil, nil, err
+	}
+
+	rule := new(Rule)
+	resp, err := p.client.Do(req, rule)
+	if err != nil {
+		return nil, resp, err
+	}
```

Reply:
helper methods don't really exist in go from what i've been reading (happy to learn otherwise). the code you identified already calls two functions from the sdk.go file so i am leaning more towards leaving this be mostly for code readability sake. i am interested in seeing if there's a way to abstract out the return statements which may be helpful. on my radar.

for this version of the file, i think the main issue is i had to add a separate struct for the signon policy as it does not support a settings field. i am unable to make that field a pointer the way the struct is designed at the moment. if i can remove that signon struct i can reduce some of the duplicate methods.

at this time, i have separate structs for policy & rule that do not contain pointers which is solely used to return data from a get call. i am thinking leaving those in will reduce code complexity for folks who use this sdk as they wont have to worry about pointers and dereferencing when they just want to query okta for api info.